File: src/org/jruby/JRubyC.java
/*
 * JRubyC.java
 *
 * Created on January 11, 2007, 11:24 PM
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

package org.jruby;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;

import java.nio.charset.CharsetDecoder;
import org.jruby.ast.Node;
import org.jruby.compiler.ASTInspector;
import org.jruby.compiler.ASTCompiler;
import org.jruby.compiler.NotCompilableException;
import org.jruby.compiler.impl.StandardASMCompiler;
import org.jruby.util.KCode;

/**
 *
 * @author headius
 */
public class JRubyC {
    
    public static void main(String args[]) {
        Ruby runtime = Ruby.newInstance();
        
        try {
            if (args.length < 1) {
                System.out.println("Usage: jrubyc <filename> [<filename> ...]");
                return;
            }
            for (int i = 0; i < args.length; i++) {
                String filename = args[i];
                if (filename.startsWith("./")) filename = filename.substring(2);
                File srcfile = new File(filename);
                if (!srcfile.exists()) {
                    System.out.println("Error -- file not found: " + filename);
                    return;
                }
                
                // destination directory
                File destfile = new File(System.getProperty("user.dir"));

                int size = (int)srcfile.length();
                byte[] chars = new byte[size];
                new FileInputStream(srcfile).read(chars);
                // FIXME: -K encoding?
                String content = new String(chars);
                Node scriptNode = runtime.parseFile(new ByteArrayInputStream(content.getBytes("ISO-8859-1")), filename, null);

                ASTInspector inspector = new ASTInspector();
                inspector.inspect(scriptNode);

                // do the compile
                String classPath = filename.substring(0, filename.lastIndexOf(".")).replace('-', '_').replace('.', '_');
                int lastSlashIndex = classPath.lastIndexOf('/');
                if (!Character.isJavaIdentifierStart(classPath.charAt(lastSlashIndex + 1))) {
                    if (lastSlashIndex == -1) {
                        classPath = "_" + classPath;
                    } else {
                        classPath = classPath.substring(0, lastSlashIndex + 1) + "_" + classPath.substring(lastSlashIndex + 1);
                    }
                }
                StandardASMCompiler compiler = new StandardASMCompiler(classPath, filename);
                ASTCompiler.compileRoot(scriptNode, compiler, inspector);

                compiler.writeClass(destfile);
            }
        } catch (IOException ioe) {
            ioe.printStackTrace();
            System.err.println("Error -- IO exception during compile: " + ioe.getMessage());
        } catch (NotCompilableException nce) {
            System.err.println("Error -- Not compilable: " + nce.getMessage());
        }
    }
    
}

File: lib/ruby/site_ruby/1.8/jruby/jrubyc.rb
require 'optparse'
require 'jruby'

module JRubyCompiler
  BAIS = java.io.ByteArrayInputStream
  Mangler = org.jruby.util.JavaNameMangler
  BytecodeCompiler = org.jruby.compiler.impl.StandardASMCompiler
  ASTCompiler = org.jruby.compiler.ASTCompiler
  JavaFile = java.io.File

  module_function
  def compile_files(filenames, basedir = Dir.pwd, prefix = "ruby", target = Dir.pwd)
    runtime = JRuby.runtime
    
    unless File.exist? target
      puts "Target dir not found: #{target}"
      exit 1
    end

    # The compilation code
    compile_proc = proc do |filename|
      begin
        file = File.open(filename)

        classpath = Mangler.mangle_filename_for_classpath(filename, basedir, prefix)
        puts "Compiling #{filename} to class #{classpath}"

        inspector = org.jruby.compiler.ASTInspector.new

        source = file.read
        node = runtime.parse_file(BAIS.new(source.to_java_bytes), filename, nil)

        inspector.inspect(node)

        asmCompiler = BytecodeCompiler.new(classpath, filename)
        compiler = ASTCompiler.new
        compiler.compile_root(node, asmCompiler, inspector)

        asmCompiler.write_class(JavaFile.new(target))
      rescue Exception
        puts "Failure during compilation of file #{filename}:\n#{$!}"
      ensure
        file.close unless file.nil?
      end
    end

    # Process all the file arguments
    filenames.each do |filename|
      unless File.exists? filename
        puts "Error -- file not found: #{filename}"
        next
      end

      if (File.directory?(filename))
        puts "Compiling all in '#{File.expand_path(filename)}'..."
        Dir.glob(filename + "/**/*.rb").each(&compile_proc)
      else
        compile_proc[filename]
      end
    end
  end
end